docs/decisions/0044-OTel-semantic-convention.md

```diff
@@ -0,0 +1,332 @@
+    var promptTokens = ...;
+    var completionTokens = ...;
+
+    ModelDiagnostics.SetCompletionResponses(
+        activity,
+        completions,
+        promptTokens,
+        completionTokens,
+        finishReasons);
+
+    return completions;
+}
+```
+
+### Switches for experimental features and the collection of sensitive data
+
+#### App context switch
+
+We will introduce two flags to facilitate the explicit activation of tracing LLMs requests:
+
+1. `Microsoft.SemanticKernel.Experimental.EnableModelDiagnostics`
+   - Activating will enable the creation of activities that represent individual LLM requests.
+2. `Microsoft.SemanticKernel.Experimental.EnableModelDiagnosticsWithSensitiveData`
+   - Activating will enable the creation of activities that represent individual LLM requests, with events that may contain PII information.
+
+```C#
+// In application code
+if (builder.Environment.IsProduction())
+{
+    AppContext.SetSwitch("Microsoft.SemanticKernel.Experimental.EnableModelDiagnostics", true);
+}
+else
+{
+    AppContext.SetSwitch("Microsoft.SemanticKernel.Experimental.EnableModelDiagnosticsWithSensitiveData", true);
+}
+
+// Or in the project file
+<ItemGroup Condition="'$(Configuration)' == 'Release'">
+    <RuntimeHostConfigurationOption Include="Microsoft.SemanticKernel.Experimental.EnableModelDiagnostics" Value="true" />
+</ItemGroup>
+
+<ItemGroup Condition="'$(Configuration)' == 'Debug'">
+    <RuntimeHostConfigurationOption Include="Microsoft.SemanticKernel.Experimental.EnableModelDiagnosticsWithSensitiveData" Value="true" />
+</ItemGroup>
+```
+
+## Decision Outcome
+
+Chosen options:
+
+[x] Scope of Activities: **Option 3** - All connectors, noting that the attributes of activities derived from connectors and those from instrumented client SDKs do not overlap.
+
+[x] Instrumentation Implementation: **Option 1** - Static class
+
+[x] Experimental switch: **Option 1** - App context switch
+
+## Appendix
+
+### `AppContextSwitchHelper.cs`
+
+```C#
+internal static class AppContextSwitchHelper
+{
+    public static bool GetConfigValue(string appContextSwitchName)
+    {
+        if (AppContext.TryGetSwitch(appContextSwitchName, out bool value))
+        {
+            return value;
+        }
+
+        return false;
+    }
+}
+```
+
+### `ModelDiagnostics`
+
+```C#
+internal static class ModelDiagnostics
+{
+    // Consistent namespace for all connectors
+    private static readonly string s_namespace = typeof(ModelDiagnostics).Namespace;
+    private static readonly ActivitySource s_activitySource = new(s_namespace);
+
+    private const string EnableModelDiagnosticsSettingName = "Microsoft.SemanticKernel.Experimental.EnableModelDiagnostics";
+    private const string EnableSensitiveEventsSettingName = "Microsoft.SemanticKernel.Experimental.EnableModelDiagnosticsWithSensitiveData";
+
+    private static readonly bool s_enableSensitiveEvents = AppContextSwitchHelper.GetConfigValue(EnableSensitiveEventsSettingName);
```

Comment:
Should we not respect changes to this AppContext switches that are made at runtime with a call to `AppContext.SetSwitch`?

Reply:
FWIW, we generally don't in dotnet/runtime. For example, we use a switch to determine whether to stick with ICU as the globalization library or to instead fall back to using NLS on Windows, and that's a choice we make once on first use and cache. I think it's fine to do that here, as well. Especially since we're talking about opting in to sharing sensitive information: this is a process-global flag, so if you were trying to do it around just an individual operation, you'd end up also enabling it for any other work in the process happening concurrently. When we want to enable this for real with public API, I think it'll make sense to do so with more fine-grained opt-in mechanisms on Kernel, which will let you do it for individual operations without concurrency concerns, but for now a global experimental switch that you just configure once should be good enough.